### PR TITLE
[Consensus] nTimeOffset warning for time protocol V2

### PR DESCRIPTION
**Pull request by random-zebra**
Open for brainstorming.
Follow up to #1002 

- reject connection with a node having offset higher than 30 secs (and don't add the sample to timedata)
- repeatedly trigger the warning message when the total offset (absolute value) is above 15 seconds (and set the offset to +15/-15 instead of 0).

### CODE REVIEW

File: src/timedata.cpp [hunk 1]
```diff
@@ -3,16 +3,14 @@
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
-#include "timedata.h"
```
Comment (Fuzzbawls) [inline src/timedata.cpp]: lets keep `timedata.h` include at the top.

also, need to include `chainparams.h` to pickup `Params()` and appease CMake

File: src/main.cpp [hunk 2]
```diff
@@ -6370,7 +6368,15 @@ bool static ProcessMessage(CNode* pfrom, std::string strCommand, CDataStream& vR
...
         int64_t nTimeOffset = nTime - GetTime();
         pfrom->nTimeOffset = nTimeOffset;
-        AddTimeData(pfrom->addr, nTimeOffset);
+        const int nTimeSlotLength = Params().TimeSlotLength();
+        if (nTimeOffset < 2 * nTimeSlotLength) {
+            pfrom->fSuccessfullyConnected = true;
+            AddTimeData(pfrom->addr, nTimeOffset, nTimeSlotLength);
+        } else {
+            LogPrintf("timeOffset (%d seconds) too large (max is %d seconds). Disconnecting node %s\n",
+                    nTimeOffset, 2 * Params().TimeSlotLength() - 1, pfrom->addr.ToString());
```
Comment (Fuzzbawls) [inline src/main.cpp]: Nit: could simplify this as
```cpp
nTimeOffset, 2 * nTimeSlotLength - 1, pfrom->addr.ToString());
```
to have one less call to `Params().TimeSlotLength()`

Comment (random-zebra) [reply src/main.cpp]: yup. nice catch

**Comment by random-zebra** [comment]
Fixed duplicate call to `Params().TimeSlotLength()` as per review.
Removing WIP label.

**Comment by random-zebra** [comment]
Indeed it is (when the clock is out of whack by more than 30 seconds with any other node on the network). User would need to check the logs in that case. Up for suggestions.

**Comment by Warrows** [comment]
Using `uiInterface.ThreadSafeMessageBox();` is more "in your face" than a warning in the logs. So it doesn't really make much sense to throw it when the clock is desynchronized by 15~30 sec but not when it is by more than 30 sec. While the opposite would make sense.
We should probably add such a call in `ProcessMessage` when no peer is connected.

**Comment by random-zebra** [comment]
The logic there is a bit more subtle.
It gets fired when it's clear that it's a node issue, because you have enough time samples.

You can't fire it in main, when disconnecting a node, because you are not sure who is "wrong" there (the node's clock might be fine... and the peer being disconnected might be the one having the clock out of whack).

**Comment by Warrows** [comment]
Maybe we should add the time sample even when we disconnect the node then.

**Comment by random-zebra** [comment]
I'm not sure this would be good either. As we would be modifying a node's offset (thus its adjustedTime) for peers not connected (possibly to anyone). Also the `vTimeOffsets` structure has room for only 200 samples. Also check bitcoin issue 4521.

**Comment by Warrows** [comment]
True, we don't want to open up clock attacks. But having the node refuse all connection with no information on what's happening outside of logs is going to get users wondering.
What about keeping a list or count (separate from time samples) and after seeing `n` nodes with differing clocks and still no connected peer fire the GUI warning from main?

**Comment by random-zebra** [comment]
Yes, this might be doable.

**Comment by Fuzzbawls** [comment]
We can address @Warrows concern in a separate PR

**Pull request merged**